internal/mode/static/nginx/modules/src/httpmatches.js

```diff
@@ -1,30 +1,55 @@
-import qs from 'querystring';
-
-const MATCHES_VARIABLE = 'http_matches';
+const MATCHES_KEY = 'match_key';
 const HTTP_CODES = {
 	notFound: 404,
 	internalServerError: 500,
 };
 
 function redirect(r) {
-	let matches;
-
+	let matchList;
 	try {
-		matches = extractMatchesFromRequest(r);
+		matchList = extractMatchesFromRequest(r, matches);
 	} catch (e) {
 		r.error(e.message);
 		r.return(HTTP_CODES.internalServerError);
 		return;
 	}
+	redirectForMatchList(r, matchList);
+}
+
+function extractMatchesFromRequest(r, matches) {
+	let matchList;
+	if (!r.variables[MATCHES_KEY]) {
+		throw Error(
+			`cannot redirect the request; the ${MATCHES_KEY} is not defined on the request object`,
+		);
+	}
 
-	// Matches is a list of http matches in order of precedence.
+	let key = r.variables[MATCHES_KEY];
+	if (!matches[key]) {
+		throw Error(
+			`cannot redirect the request; the key ${key} is not defined on the matches object`,
+		);
+	}
+
+	// matchList is a list of http matches in order of precedence.
 	// We will accept the first match that the request satisfies.
 	// If there's a match, redirect request to internal location block.
 	// If an exception occurs, return 500.
 	// If no matches are found, return 404.
+	matchList = matches[key];
+	try {
+		verifyMatchList(matchList);
+	} catch (e) {
+		throw Error(`cannot redirect the request; ${matchList} matches list is not valid`);
```

Comment:
let's make sure we use the error message from e in the exception we through here